File: du-backend/src/main/java/com/dwin/du/api/service/ChartService.java
package com.dwin.du.api.service;
import com.dwin.du.api.dto.ChartDto;
import com.dwin.du.api.entity.Person;
import com.dwin.du.api.entity.PersonProduct;
import com.dwin.du.api.entity.Receipt;
import com.dwin.du.api.repository.PersonProductRepository;
import com.dwin.du.api.repository.PersonRepository;
import com.dwin.du.api.entity.User;
import com.dwin.du.api.repository.ReceiptRepository;
import com.dwin.du.validation.ValidationException;
import com.dwin.du.validation.ValidationService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ChartService {

    private final PersonRepository personRepository;
    private final ReceiptRepository receiptRepository;
    private final PersonProductRepository personProductRepository;
    private final ValidationService validator;

    private static final String[] MONTH_NAMES = {"Sty", "Lut", "Mar", "Kwi", "Maj", "Cze", "Lip", "Sie", "Wrz", "Paź", "Lis", "Gru"};
    private static final String[] WEEK_NAMES = {"Pon", "Wt", "Śr", "Czw", "Pt", "Sb", "Ndz"};

    public ResponseEntity<?> getChartAmounts(String username, boolean isTotalAmounts) {
        try {
            User user = validator.validateUser(username);
            List<Person> persons = personRepository.findByUser(user);

            List<ChartDto> responseList = new ArrayList<>();
            for (Person person : persons) {
                double amount = isTotalAmounts ? person.getTotalAmount() : person.getUnpaidAmount();
                if (Double.isNaN(amount))
                    amount = 0.0;
                //amount = Math.round(amount * 100.0) / 100.0;

                ChartDto response = ChartDto.builder()
                        .name(person.getName()+" "+person.getSurname())
                        .value(amount)
                        .build();
                responseList.add(response);
            }
            return ResponseEntity.ok(responseList);

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> getPercentageExpanses(String username) {
        try {
            User user = validator.validateUser(username);
            List<Person> persons = personRepository.findByUser(user);

            List<ChartDto> responseList = new ArrayList<>();
            for (Person person : persons) {
                double amount = calculatePaidPercentage(person.getTotalAmount(),person.getUnpaidAmount());

                ChartDto response = ChartDto.builder()
                        .name(person.getName()+" "+person.getSurname())
                        .value(amount)
                        .build();
                responseList.add(response);
            }
            return ResponseEntity.ok(responseList);

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> getMonthlyTotalExpenses(String username, int year) {
        try {
            User user = validator.validateUser(username);

            List<Receipt> receipts = receiptRepository.findByUser(user);
            Map<Integer, Double> monthlyTotals = new HashMap<>();

            for (Receipt receipt : receipts) {
                LocalDate date = receipt.getDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate();

                if (date.getYear() == year) {
                    int month = date.getMonthValue();
                    monthlyTotals.put(month, monthlyTotals.getOrDefault(month, 0.0) + receipt.getTotalPrice());
                }
            }

            List<ChartDto> responseList = new ArrayList<>();
            for (int month = 1; month <= 12; month++) {
                double total = monthlyTotals.getOrDefault(month, 0.0);
                responseList.add(ChartDto.builder()
                        .name(MONTH_NAMES[month - 1])
                        .value(total)
                        .build());
            }

            return ResponseEntity.ok(responseList);

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> getMonthlyUserExpenses(String username, int year) {
        try {
            User user = validator.validateUser(username);

            List<Person> persons = personRepository.findByUser(user);
            List<Person> userAccountPersons = persons.stream()
                    .filter(Person::isUserAccount)
                    .toList();

            Map<Integer, Double> monthlyTotals = new HashMap<>();

            for (Person person : userAccountPersons) {
                List<PersonProduct> personProducts = personProductRepository.findByPerson(person);

                for (PersonProduct personProduct : personProducts) {
                    LocalDate date = personProduct.getProduct().getReceipt().getDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate();

                    if (date.getYear() == year) {
                        int month = date.getMonthValue();
                        monthlyTotals.put(month, monthlyTotals.getOrDefault(month, 0.0) + personProduct.getPartOfPrice());
                    }
                }
            }

            List<ChartDto> responseList = new ArrayList<>();
            for (int month = 1; month <= 12; month++) {
                double total = monthlyTotals.getOrDefault(month, 0.0);
                responseList.add(ChartDto.builder()
                        .name(MONTH_NAMES[month - 1])
                        .value(total)
                        .build());
            }

            return ResponseEntity.ok(responseList);

        } catch (Exception e) {
            return handleException(e);
        }
    }

    private double calculatePaidPercentage(double totalAmount, double unpaidAmount) {
        if (totalAmount == 0) return 100;
        if (unpaidAmount == 0) return 100;
        double percentage = (1 - (unpaidAmount / totalAmount)) * 100;
        if (Double.isNaN(percentage))
            percentage = 0.0;

        return percentage;
        //return Math.round(percentage * 100.0) / 100.0;
    }


    public ResponseEntity<?> getWeeklyTotalExpenses(String username) {
        try {
            User user = validator.validateUser(username);
            List<Receipt> receipts = receiptRepository.findByUser(user);

            Map<DayOfWeek, Double> dailyTotals = new HashMap<>();
            for (DayOfWeek day : DayOfWeek.values())
                dailyTotals.put(day, 0.0);

            LocalDate today = LocalDate.now();
            LocalDate startOfWeek = today.with(DayOfWeek.MONDAY);
            LocalDate endOfWeek = today.with(DayOfWeek.SUNDAY);

            for (Receipt receipt : receipts) {
                LocalDate date = receipt.getDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate();

                if (!date.isBefore(startOfWeek) && !date.isAfter(endOfWeek)) {
                    DayOfWeek dayOfWeek = date.getDayOfWeek();
                    dailyTotals.put(dayOfWeek, dailyTotals.get(dayOfWeek) + receipt.getTotalPrice());
                }
            }

            List<ChartDto> responseList = new ArrayList<>();
            for (DayOfWeek day : DayOfWeek.values()) {
                double total = dailyTotals.getOrDefault(day, 0.0);
                if (Double.isNaN(total))
                    total = 0.0;

                responseList.add(ChartDto.builder()
                        .name(WEEK_NAMES[day.getValue() - 1])
                        .value(total)
                        .build());
            }


            return ResponseEntity.ok(responseList);

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> getWeeklyUserExpenses(String username) {
        try {
            User user = validator.validateUser(username);
            List<Person> persons = personRepository.findByUser(user);
            List<Person> userAccountPersons = persons.stream()
                    .filter(Person::isUserAccount)
                    .toList();

            Map<DayOfWeek, Double> dailyTotals = new HashMap<>();
            for (DayOfWeek day : DayOfWeek.values())
                dailyTotals.put(day, 0.0);

            LocalDate today = LocalDate.now();
            LocalDate startOfWeek = today.with(DayOfWeek.MONDAY);
            LocalDate endOfWeek = today.with(DayOfWeek.SUNDAY);

            for (Person person : userAccountPersons) {
                List<PersonProduct> personProducts = personProductRepository.findByPerson(person);

                for (PersonProduct personProduct : personProducts) {
                    LocalDate date = personProduct.getProduct().getReceipt().getDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate();

                    if (!date.isBefore(startOfWeek) && !date.isAfter(endOfWeek)) {
                        DayOfWeek dayOfWeek = date.getDayOfWeek();
                        dailyTotals.put(dayOfWeek, dailyTotals.get(dayOfWeek) + personProduct.getPartOfPrice());
                    }
                }
            }

            List<ChartDto> responseList = new ArrayList<>();
            for (DayOfWeek day : DayOfWeek.values()) {
                double total = dailyTotals.getOrDefault(day, 0.0);
                responseList.add(ChartDto.builder()
                        .name(WEEK_NAMES[day.getValue() - 1])
                        .value(total)
                        .build());
            }

            return ResponseEntity.ok(responseList);

        } catch (Exception e) {
            return handleException(e);
        }
    }

    public ResponseEntity<?> getMonthlyTopProducts(String username) {
        try {
            User user = validator.validateUser(username);
            List<Person> persons = personRepository.findByUser(user);

            List<PersonProduct> allPersonProducts = new ArrayList<>();
            for (Person person : persons) {
                allPersonProducts.addAll(personProductRepository.findByPerson(person));
            }

            LocalDate now = LocalDate.now();
            int currentMonth = now.getMonthValue();
            int currentYear = now.getYear();

            Map<String, Double> productValues = new HashMap<>();

            for (PersonProduct personProduct : allPersonProducts) {
                LocalDate productDate = personProduct.getProduct().getReceipt().getDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
                if (productDate.getMonthValue() == currentMonth && productDate.getYear() == currentYear) {
                    productValues.merge(personProduct.getProduct().getName(), personProduct.getPartOfPrice(), Double::sum);
                }
            }

            List<ChartDto> topProducts = productValues.entrySet().stream()
                    .sorted(Map.Entry.<String, Double>comparingByValue().reversed())
                    .limit(3)
                    .map(entry -> ChartDto.builder()
                            .name(entry.getKey())
                            .value(entry.getValue())
                            .build())
                    .collect(Collectors.toList());

            return ResponseEntity.ok(topProducts);

        } catch (Exception e) {
            return handleException(e);
        }
    }


    private ResponseEntity<?> handleException(Exception e) {
        if (e instanceof ValidationException) {
            HttpStatus status = ((ValidationException) e).getStatus();
            return ResponseEntity.status(status).body(e.getMessage());
        }
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Wystąpił nieoczekiwany błąd.");
    }
}

File: divvyup/DivvyUp.Web/Services/ChartService.cs
using DivvyUp.Web.Data;
using DivvyUp.Web.EntityManager;
using DivvyUp_Shared.Dtos.Entity;
using DivvyUp_Shared.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace DivvyUp.Web.Services
{
    public class ChartService : IChartService
    {
        private readonly DivvyUpDBContext _dbContext;
        private readonly EntityManagementService _managementService;

        private static readonly string[] MonthNames = { "Sty", "Lut", "Mar", "Kwi", "Maj", "Cze", "Lip", "Sie", "Wrz", "Paź", "Lis", "Gru" };
        private static readonly string[] WeekNames = { "Niedz", "Pon", "Wt", "Śr", "Czw", "Pt", "Sb" };

        public ChartService(DivvyUpDBContext dbContext, EntityManagementService managementService)
        {
            _dbContext = dbContext;
            _managementService = managementService;
        }

        public async Task<List<ChartDto>> GetAmounts(bool isTotalAmounts)
        {
            var user = await _managementService.GetUser();
            var persons = await _dbContext.Persons.Where(p => p.UserId == user.Id).ToListAsync();

            var responseList = new List<ChartDto>();
            foreach (var person in persons)
            {
                decimal amount = isTotalAmounts ? person.TotalAmount : person.UnpaidAmount;

                var response = new ChartDto
                {
                    Name = $"{person.Name} {person.Surname}",
                    Value = amount
                };
                responseList.Add(response);
            }

            return responseList;
        }

        public async Task<List<ChartDto>> GetPercentageExpanses()
        {
            var user = await _managementService.GetUser();
            var persons = await _dbContext.Persons
                .AsNoTracking()
                .Where(p => p.UserId == user.Id).ToListAsync();

            var responseList = new List<ChartDto>();
            foreach (var person in persons)
            {
                decimal totalAmount = person.TotalAmount + (person.LoanBalance < 0 ? person.LoanBalance * -1 : 0);
                decimal unpaidAmount = person.UnpaidAmount + (person.LoanBalance < 0 ? person.LoanBalance * -1 : 0);

                decimal reliabilityPercentage = CalculatePaidPercentage(totalAmount, unpaidAmount);

                var response = new ChartDto
                {
                    Name = $"{person.Name} {person.Surname}",
                    Value = reliabilityPercentage
                };
                responseList.Add(response);
            }
            return responseList;
        }

        public async Task<List<ChartDto>> GetMonthlyTotalExpenses(int year)
        {

            var user = await _managementService.GetUser();

            var receipts = await _dbContext.Receipts
                .AsNoTracking()
                .Where(r => r.UserId == user.Id && r.Date.Year == year)
                .ToListAsync();

            var monthlyTotals = new Dictionary<int, decimal>();

            foreach (var receipt in receipts)
            {
                int month = receipt.Date.Month;
                monthlyTotals[month] = monthlyTotals.GetValueOrDefault(month, 0.0m) + receipt.TotalPrice;
            }

            var responseList = new List<ChartDto>();

            for (int month = 1; month <= 12; month++)
            {
                decimal total = monthlyTotals.GetValueOrDefault(month, 0.0m);
                responseList.Add(new ChartDto
                {
                    Name = MonthNames[month - 1],
                    Value = total
                });
            }
            return responseList;
        }

        public async Task<List<ChartDto>> GetMonthlyUserExpenses(int year)
        {
            var user = await _managementService.GetUser();
            var persons = await _dbContext.Persons
                .AsNoTracking()
                .Where(p => p.UserId == user.Id)
                .ToListAsync();
            var userAccountPersons = persons.Where(p => p.UserAccount).ToList();

            var monthlyTotals = new Dictionary<int, decimal>();

            foreach (var person in userAccountPersons)
            {
                var personProducts = await _dbContext.PersonProducts
                    .AsNoTracking()
                    .Include(pp => pp.Product)
                    .Include(pp => pp.Product.Receipt)
                    .Where(pp => pp.PersonId == person.Id)
                    .ToListAsync();


                foreach (var personProduct in personProducts)
                {
                    if (personProduct?.Product?.Receipt != null)
                    {
                        var date = personProduct.Product.Receipt.Date;
                        if (date.Year == year)
                        {
                            int month = date.Month;
                            monthlyTotals[month] = monthlyTotals.GetValueOrDefault(month, 0.0m) + personProduct.PartOfPrice;
                        }
                    }
                }
            }

            var responseList = new List<ChartDto>();
            for (int month = 1; month <= 12; month++)
            {
                decimal total = monthlyTotals.GetValueOrDefault(month, 0.0m);
                responseList.Add(new ChartDto
                {
                    Name = MonthNames[month - 1],
                    Value = total
                });
            }

            return responseList;
        }

        public async Task<List<ChartDto>> GetWeeklyTotalExpenses()
        {
            var today = DateOnly.FromDateTime(DateTime.Today);
            var startOfWeek = today.AddDays(-((int)today.DayOfWeek - (int)DayOfWeek.Monday + 7) % 7);
            var endOfWeek = startOfWeek.AddDays(6);

            var user = await _managementService.GetUser();

            var receipts = await _dbContext.Receipts
                .AsNoTracking()
                .Where(r => r.UserId == user.Id && r.Date >= startOfWeek && r.Date <= endOfWeek)
                .ToListAsync();

            var dailyTotals = new Dictionary<DayOfWeek, decimal>();

            foreach (var receipt in receipts)
            {
                var dayOfWeek = receipt.Date.DayOfWeek;
                dailyTotals[dayOfWeek] = dailyTotals.GetValueOrDefault(dayOfWeek, 0.0m) + receipt.TotalPrice;
            }

            var responseList = new List<ChartDto>();

            for (int i = 0; i < 7; i++)
            {
                var currentDay = startOfWeek.AddDays(i);
                var total = dailyTotals.GetValueOrDefault(currentDay.DayOfWeek, 0.0m);
                responseList.Add(new ChartDto
                {
                    Name = WeekNames[(int)currentDay.DayOfWeek],
                    Value = total
                });
            }

            return responseList;
        }



        public async Task<List<ChartDto>> GetWeeklyUserExpenses()
        {
            var today = DateOnly.FromDateTime(DateTime.Today);
            var startOfWeek = today.AddDays(-((int)today.DayOfWeek - (int)DayOfWeek.Monday + 7) % 7);
            var endOfWeek = startOfWeek.AddDays(6);

            var user = await _managementService.GetUser();

            var persons = await _dbContext.Persons
                .AsNoTracking()
                .Where(p => p.UserId == user.Id)
                .ToListAsync();
            var userAccountPersons = persons.Where(p => p.UserAccount).ToList();

            var dailyTotals = new Dictionary<DayOfWeek, decimal>();

            foreach (var person in userAccountPersons)
            {
                var personProducts = await _dbContext.PersonProducts
                    .AsNoTracking()
                    .Include(pp => pp.Product)
                    .Include(pp => pp.Product.Receipt)
                    .Where(pp => pp.PersonId == person.Id)
                    .ToListAsync();

                foreach (var personProduct in personProducts)
                {
                    if (personProduct?.Product?.Receipt != null)
                    {
                        var date = personProduct.Product.Receipt.Date;
                        if (date >= startOfWeek && date <= endOfWeek)
                        {
                            var dayOfWeek = date.DayOfWeek;
                            dailyTotals[dayOfWeek] = dailyTotals.GetValueOrDefault(dayOfWeek, 0.0m) + personProduct.PartOfPrice;
                        }
                    }
                }
            }

            var responseList = new List<ChartDto>();

            for (int i = 0; i < 7; i++)
            {
                var currentDay = startOfWeek.AddDays(i);
                var total = dailyTotals.GetValueOrDefault(currentDay.DayOfWeek, 0.0m);
                responseList.Add(new ChartDto
                {
                    Name = WeekNames[(int)currentDay.DayOfWeek],
                    Value = total
                });
            }

            return responseList;
        }

        public async Task<List<ChartDto>> GetMonthlyTopProducts()
        {
            var user = await _managementService.GetUser();

            var now = DateTime.Now;
            var currentMonth = now.Month;
            var currentYear = now.Year;

            var topProducts = await _dbContext.Products
                .Include(p => p.Receipt)
                .Where(p => p.Receipt.UserId == user.Id &&
                            p.Receipt.Date.Month == currentMonth &&
                            p.Receipt.Date.Year == currentYear)
                .Select(p => new { p.Name, p.Price })
                .Distinct()
                .OrderByDescending(p => p.Price)
                .Take(3)
                .Select(p => new ChartDto(p.Name, p.Price))
                .ToListAsync();

            return topProducts;
        }

        private decimal CalculatePaidPercentage(decimal totalAmount, decimal unpaidAmount)
        {
            if (totalAmount == 0) return 100;
            if (unpaidAmount == 0) return 100;

            decimal percentage = (1 - (unpaidAmount / totalAmount)) * 100;
            return Math.Round(percentage, 2);
        }
    }
}
